shared/src/tx_pool_ext.rs

```diff
@@ -212,57 +208,71 @@ impl TxPool {
-                            .is_some()
-                        {
-                            self.update_statics_for_remove_tx(size, cycles.unwrap_or(0));
-                        }
+            Err(err) => {
+                match err.kind() {
+                    ErrorKind::Transaction => {
+                        self.update_statics_for_remove_tx(size, cycles.unwrap_or(0));
+                        debug_target!(
+                            crate::LOG_TARGET_TX_POOL,
+                            "Failed to add tx to {} {}, verify failed, reason: {:?}",
+                            pool_name,
+                            tx_hash,
+                            err,
+                        );
                     }
-                    UnresolvableError::Unknown(out_points) => {
-                        if self
-                            .add_orphan(cycles, size, tx, out_points.to_owned())
-                            .is_some()
+                    ErrorKind::OutPoint => {
+                        match err
+                            .downcast_ref::<OutPointError>()
+                            .expect("error kind checked")
                         {
-                            self.update_statics_for_remove_tx(size, cycles.unwrap_or(0));
+                            OutPointError::Dead(_) => {
+                                if self
+                                    .conflict
+                                    .insert(short_id, DefectEntry::new(tx, 0, cycles, size))
+                                    .is_some()
+                                {
+                                    self.update_statics_for_remove_tx(size, cycles.unwrap_or(0));
+                                }
+                            }
+
+                            OutPointError::Unknown(out_points) => {
+                                if self
+                                    .add_orphan(cycles, size, tx, out_points.to_owned())
+                                    .is_some()
+                                {
+                                    self.update_statics_for_remove_tx(size, cycles.unwrap_or(0));
+                                }
+                            }
+
+                            // The remaining errors represent invalid transactions that should
+                            // just be discarded.
+                            //
+                            // To avoid mis-discarding error types added in the future, please don't
+                            // use `Some(_)` as the match arm.
+                            //
+                            // OutOfOrder should only appear in BlockCellProvider
+                            OutPointError::InvalidHeaderDep(_)
+                            | OutPointError::InvalidDepGroup(_)
+                            | OutPointError::Empty(_)
+                            | OutPointError::UnspecifiedInputCell(_)
```

Comment:
When does it raise `UnspecifiedInputCell` and `UnknownHeader`?

Reply:
Oh, I didn't notice that they were removed from https://github.com/nervosnetwork/ckb/pull/1356

Thanks! I will remove the two errors.